share/spack/gitlab/cloud_pipelines/stacks/hep/spack.yaml

```diff
@@ -0,0 +1,64 @@
+spack:
+  view: false
+
+  concretizer:
+    reuse: false
+    unify: when_possible
+
+  packages:
+    all:
+      require: '%gcc target=x86_64_v3'
+      providers:
+        blas: [openblas]
+        mpi: [mpich]
+        tbb: [intel-tbb]
+      variants: +mpi
+
+    # Mark geant4 data as external to prevent wasting bandwidth on GB-scale files
+    geant4-data:
+      externals:
+        - spec: geant4-data@11.3.0
+          prefix: /usr
+        - spec: geant4-data@11.2.2
+          prefix: /usr
+        - spec: geant4-data@11.2.0
+          prefix: /usr
+        - spec: geant4-data@11.1.0
+          prefix: /usr
+        - spec: geant4-data@11.0.0
+          prefix: /usr
+
+  specs:
+  # CPU
+  - acts +analysis +dd4hep +edm4hep +examples +fatras +geant4 +hepmc3 +podio +pythia8 +python +tgeo cxxstd=20
```

Comment:
What's considered 'huge'? :-) Here's a trace for https://gitlab.spack.io/spack/spack/-/pipelines/914877

![image](https://github.com/user-attachments/assets/e5b343ae-d61d-4338-ab3e-77a58b0f3c13)

Reply:
(some downstream pkgs are still missing on that trace, but I think it's only really `acts` and `dd4hep` that takes a similar amount of time)

Reply:
Hmm, I'd need to run some benchmarking on that then. Are there tools in the gitlab runners that run sampling profiling on the build jobs to determine how many of the allocated cores they keep occupied? All I could find was the loadavg and the memory use at completion. More useful for this would be a report of core use and peak memory use.

I have to admit that I am usually running my compiles of these larger packages with `-j 128` on a dedicated build server that's overspeced on memory.

Reply:
I was playing around with prmon for some profiling, and could get some graphs of a spack install like this:
![image](https://github.com/user-attachments/assets/60daa02f-2fd1-4e5d-baeb-bc1d2d5053fc)
![image](https://github.com/user-attachments/assets/b512476c-05be-428c-811f-bae3dac48d80)
This could be useful if it was accessible for each build. In this case it was multiple builds (the larger one for 500 s at the end), but in CI this could be constrained to the single package build, I assume.

Reply:
That's pretty neat, I haven't used prmon before. We have stuff kind of like that using Prometheus for jobs, but I don't know if the data is generally available though.

For the most part a general idea of allocation is enough and we can adjust things later. I mostly just want to make sure that we aren't leaving the lower default and have to push fixes to bump the resources up for new packages while it is failing in CI for PRs.

Reply:
## Profiling results

Largely similar specs, but some variants are different due to my local `unify: true` requirement; some local recipe changes; obtained on a quiescent skylake system with `-j8`. Cleared ccache before, but didn't entirely disable it (maybe 5% overhead).

Below all builds > 5 minute, so the rest could run within 20 minutes on the default 2-core allocation.

- All large builds show good scaling at 8 threads: integrated user+system time is about 7.5 x wall time. Lower utilization during tests.
- Memory spikes at linking time.

Proposal is to use:
- "huge" group with dealii, rocblas
  - geant4: 12 cores, 12 GB, 37 minutes (extrapolated)
- "large" group with ecp-data-vis-sdk, intel-tbb, llvm-amdgpu, salmon, trilinos
  - acts: 8 cores, 15 GB, 25 minutes
- "large" group with vtk and lots of packages
  - gaudi: 8 cores, 12 GB, 10 minutes
- "large" group with hipblas, rocfft, umpire
  - rivet: 8 cores, 8 GB, 11 minutes
  - dd4hep: 8 cores, 8 GB, 12 minutes
  - qt-base: 8 cores, 8 GB, 15 minutes

### Acts: 8 cores, 16 GB, 25 minutes
```
6infqnyypnbzkyohqshsuorbice2kn7n acts@35.2.0~alignment+analysis~benchmarks~cuda+dd4hep+edm4hep+examples+fatras~fatras_geant4+geant4~geomodel+hepmc3~integration_tests~ipo+json~legacy+odd~onnx+podio+profilecpu+profilemem+pythia8+python~svg+tbb+tgeo~traccc~unit_tests build_system=cmake build_type=RelWithDebInfo cxxstd=20 dev_path=/home/wdconinc/git/acts generator=make log_failure_threshold=MAX scalar=double
```
![6infqnyypnbzkyohqshsuorbice2kn7n_time](https://github.com/user-attachments/assets/6f7fff4b-7313-453e-8392-da8dc79876a8) ![6infqnyypnbzkyohqshsuorbice2kn7n_mem](https://github.com/user-attachments/assets/fdb31e70-67e2-47d1-b8f8-37292da2dc2e)

### Gaudi: 8 cores, 12 GB, 10 minutes
```
b7zf3i4es3zej774a6mnchz664l6nikj gaudi@39.0+aida~cppunit~docs~examples~gperftools+heppdt~ipo~jemalloc~unwind~vtune+xercesc build_system=cmake build_type=Release generator=make
```
![b7zf3i4es3zej774a6mnchz664l6nikj_time](https://github.com/user-attachments/assets/4362acf0-6c61-4aec-93e1-7a50945478b0)
![b7zf3i4es3zej774a6mnchz664l6nikj_mem](https://github.com/user-attachments/assets/60603d9f-2942-41ed-a129-6914e46bf8d6)

### Rivet: 8 cores, 8 GB, 11 minutes
```
gvjlgo64t726cflfmhfweg2ldkdj3b6w rivet@3.1.10 build_system=autotools hepmc=3
```
![gvjlgo64t726cflfmhfweg2ldkdj3b6w_time](https://github.com/user-attachments/assets/40415773-00e8-49d3-aa74-dfbd77bdedbd)
![gvjlgo64t726cflfmhfweg2ldkdj3b6w_mem](https://github.com/user-attachments/assets/e26c1c59-53bf-4ba0-bb35-bf56b17907e1)

### DD4hep: 8 cores, 8 GB, 12 minutes
```
rzgn7tsalzrfuzbc3nio3gnyoeivbyc5 dd4hep@1.30+ddalign+ddcad+ddcond+dddetectors+dddigi+ddeve+ddg4+ddrec~debug+edm4hep~examples~geant4units+hepmc3~hepmc3-gz~ipo~lcio+tbb+utilityapps+xercesc build_system=cmake build_type=Release dev_path=/home/wdconinc/git/DD4hep generator=ninja patches=3858ac2,fe28edb
```
![rzgn7tsalzrfuzbc3nio3gnyoeivbyc5_time](https://github.com/user-attachments/assets/e976327f-faee-43d7-8416-6ca8ad7025ee)
![rzgn7tsalzrfuzbc3nio3gnyoeivbyc5_mem](https://github.com/user-attachments/assets/12c13350-7dbb-4d6f-8c5a-fe702e592c05)

### Qt-base: 8 cores, 8 GB, 15 minutes
```
vzza2nnqm4mulge4ecwkyjcab2nlfbi7 qt-base@6.8.0+accessibility~dbus~framework~gtk+gui~ipo+network+opengl+shared+sql+widgets build_system=cmake build_type=Release generator=ninja
```
![vzza2nnqm4mulge4ecwkyjcab2nlfbi7_time](https://github.com/user-attachments/assets/0d5d4e43-5cac-4b93-872b-db3a53ad62b6)
![vzza2nnqm4mulge4ecwkyjcab2nlfbi7_mem](https://github.com/user-attachments/assets/9a4809a0-f058-4e4d-8dc3-b52237cb59c5)

### Geant4: 8 cores, 8 GB, 56 minutes
```
y6j3lvcbo3rfesdbaebjvfb3cg6qjnpy geant4@11.3.0+data~hdf5~ipo~motif+opengl+qt~tbb+threads~timemory+vecgeom~vtk~x11 build_system=cmake build_type=RelWithDebInfo cxxstd=20 generator=make
```
![y6j3lvcbo3rfesdbaebjvfb3cg6qjnpy_time](https://github.com/user-attachments/assets/3d83e291-f4fb-43cf-8dcf-cfbfcfe934be)
![y6j3lvcbo3rfesdbaebjvfb3cg6qjnpy_mem](https://github.com/user-attachments/assets/b2cba25f-011a-44f6-aa77-2409c2f5dd16)

Reply:
Hmm, apparently my profiling results of `root` were not saved correctly so it is not in the list above but likely should be...

Reply:
### Qt-declarative: 8 cores, 8 GB, 25 minutes
```
jiibgb3nkulbocpk5z7modvdnjw74w4u qt-declarative@6.8.0~ipo build_system=cmake build_type=Release generator=ninja
```
![jiibgb3nkulbocpk5z7modvdnjw74w4u_time](https://github.com/user-attachments/assets/3beca80e-2188-45b3-a307-1d6c52931ac2)
![jiibgb3nkulbocpk5z7modvdnjw74w4u_mem](https://github.com/user-attachments/assets/b4dcefe0-0acd-49e9-8653-bee772ff10cf)

Reply:
### Sherpa: 8core, 3 GB, 13 minutes
```
joxgkhdzu3zh6rbrj2zaj7t2ehgwqgfq sherpa@3.0.1+analysis~blackhat~cms~gzip+hepmc3~hepmc3root~ipo+lhapdf~lhole~mpi~openloops+pythia+python~recola+rivet~root~ufo build_system=cmake build_type=Release cxxstd=14 generator=make libs=shared,static
```
![joxgkhdzu3zh6rbrj2zaj7t2ehgwqgfq_time](https://github.com/user-attachments/assets/8ed1946f-f03d-4f28-9b05-fb5059a8d620)
![joxgkhdzu3zh6rbrj2zaj7t2ehgwqgfq_mem](https://github.com/user-attachments/assets/9e2e85c4-eff7-4295-bd4b-153ab7ec304a)

Reply:
### ROOT: 8 cores, 9 GB, 110 minutes
```
ngm7adwbwhqnc3ckcnrhx2mbqbqp447v root@6.32.08~aqua~arrow~cuda~cudnn+davix+dcache~emacs+examples+fftw+fits~fortran+gdml+gminimal+graphviz+gsl+http~ipo+math+minuit+mlp~mysql+opengl~postgres~pythia8+python+r+roofit+root7+rpath~shadow+spectrum+sqlite+ssl+tbb+threads+tmva+tmva-cpu~tmva-gpu~tmva-pymva~tmva-sofie+unuran+vc+vdt+veccore+webgui+x+xml+xrootd build_system=cmake build_type=Release cxxstd=20 generator=make patches=22af347
```
![ngm7adwbwhqnc3ckcnrhx2mbqbqp447v_time](https://github.com/user-attachments/assets/34ce2c6a-e066-4ebb-80a1-eacf96ff0d25)
![ngm7adwbwhqnc3ckcnrhx2mbqbqp447v_mem](https://github.com/user-attachments/assets/8f56e4c6-7f93-4ca4-b889-16f143820168)

Proposal to add to huge as well.